gpkit/tools.py

```diff
@@ -112,3 +115,53 @@ def mdmake(filename, make_tex=True):
         with open(filename+".tex.md", "w") as f:
             f.write(texmd)
     return open(filename+".py")
+
+
+def bound_all_variables(model, eps=1e-30, lower=None, upper=None):
+    "Returns model with additional constraints bounding all free variables"
+    lb = lower if lower else eps
+    ub = upper if upper else 1/eps
+    constraints = []
+    for varkey in model.varkeys:
+        units = varkey.descr.get("units", 1)
+        constraints.append([ub*units >= Variable(**varkey.descr),
+                            Variable(**varkey.descr) >= lb*units])
+    m = Model(model.cost, [constraints, model], model.substitutions)
+    m.bound_all = {"lb": lb, "ub": ub, "varkeys": model.varkeys}
+    return m
+
+
+# pylint: disable=too-many-locals
+def determine_unbounded_variables(model, solver=None, verbosity=0,
+                                  eps=1e-30, lower=None, upper=None, **kwargs):
+    "Returns labeled dictionary of unbounded variables."
+    m = bound_all_variables(model, eps, lower, upper)
+    sol = m.solve(solver, verbosity, **kwargs)
+    lam = sol["sensitivities"]["la"][1:]
+    out = defaultdict(list)
+    for i, varkey in enumerate(m.bound_all["varkeys"]):
+        lam_gt, lam_lt = lam[2*i], lam[2*i+1]
+        if lam_gt and lam_lt:  # to avoid divide-by-zero
+            sens_ratio = lam_gt/lam_lt
+        elif lam_gt:
+            sens_ratio = np.inf
+        else:
+            sens_ratio = 0.0
+        if sens_ratio >= 2:  # arbitrary threshold
+            out["pushing up"].append(varkey)
+        elif sens_ratio <= 0.5:  # arbitrary threshold
+            out["pushing down"].append(varkey)
+        else:
+            value = mag(sol["variables"][varkey])
+            distance_below = np.log(value/m.bound_all["lb"])
+            distance_above = np.log(m.bound_all["ub"]/value)
+            if distance_below <= 3:
+                out["hit eps lower bound"].append(varkey)
+            elif distance_above <= 3:
+                out["hit 1/eps upper bound"].append(varkey)
+           #  if distance_below <= 3 or distance_above <= 3:
+           #      # if we're within ~10x of a boundary (arbitrary threshold)
+           #      raise AttributeError("%s appears to have hit a boundary"
+           #                           " but it's not showing up in the"
+           #                           " sensitivities." % varkey)
```

Comment:
@pgkirsch and I are still discussing what constitutes a boundary "hit":
just sensitivities, or sensitivities and position?
On Mar 25, 2016 12:36, "Warren Hoburg" <notifications@github.com> wrote:

> In gpkit/tools.py
> <https://github.com/hoburg/gpkit/pull/571#discussion_r57461159>:
>
> > +            out["pushing up"].append(varkey)
> > +        elif sens_ratio <= 0.5:  # arbitrary threshold
> > +            out["pushing down"].append(varkey)
> > +        else:
> > +            value = mag(sol["variables"][varkey])
> > +            distance_below = np.log(value/m.bound_all["lb"])
> > +            distance_above = np.log(m.bound_all["ub"]/value)
> > +            if distance_below <= 3:
> > +                out["hit eps lower bound"].append(varkey)
> > +            elif distance_above <= 3:
> > +                out["hit 1/eps upper bound"].append(varkey)
> > +           #  if distance_below <= 3 or distance_above <= 3:
> > +           #      # if we're within ~10x of a boundary (arbitrary threshold)
> > +           #      raise AttributeError("%s appears to have hit a boundary"
> > +           #                           " but it's not showing up in the"
> > +           #                           " sensitivities." % varkey)
>
> should this commented out code stay?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/hoburg/gpkit/pull/571/files/df5a416930d67c14369ba442ac71ae614255fc53#r57461159>
>